crates/parser/src/grammar.rs

```diff
@@ -185,9 +185,14 @@ fn opt_visibility(p: &mut Parser) -> bool {
                     // pub(self) struct S;
                     // pub(self) struct S;
                     T![crate] | T![self] | T![super] => {
-                        p.bump_any();
-                        p.bump_any();
-                        p.expect(T![')']);
+                        // test pub_parens_typepath
+                        // struct B(pub (super::A));
+                        // struct B(pub (crate::A,));
+                        if p.nth(2) != T![:] {
```

Comment:
Let's move this condition to the match arm guard.

Bonus points: fix **Move condition to match guard** assists such that not only this works:

![image](https://user-images.githubusercontent.com/1711539/97057477-976be480-158b-11eb-9d74-d43b026f0d0f.png)


But this as well:

![image](https://user-images.githubusercontent.com/1711539/97057500-a3f03d00-158b-11eb-857b-767d31dd17f4.png)

Reply:
Don't merge yet.
While I was eating my own dog food,
![dog_food1](https://user-images.githubusercontent.com/23637819/97069815-c6d91c00-15a0-11eb-8772-1c7aa81eed55.png)
![dog_food2](https://user-images.githubusercontent.com/23637819/97069817-c93b7600-15a0-11eb-9731-aa007f9d9938.png)
 I realized that comments sometimes get eaten:
```rust
fn main() {
    match 92 {
        x => {
            // This comment gets eaten
            if x > 10 {
                92;
                // This comment survives
                false
            }
        }
        _ => true
    }
}
```
Will fix tomorrow.

Reply:
How should comments in blocks be handled? Does the behavior of these tests make sense?

<details>
<summary>Tests</summary>

```rust
    fn move_arm_cond_with_comments_to_match_guard_works() {
        check_assist(
            move_arm_cond_to_match_guard,
            r#"
fn main() {
    match 92 {
        // Comment A
        x => if x > 10 {<|>
            // Comment B
            false
        },
        _ => true
    }
}
"#,
            r#"
fn main() {
    match 92 {
        // Comment A
        // Comment B
        x if x > 10 => false,
        _ => true
    }
}
"#,
        );
    }

    #[test]
    fn move_arm_cond_with_comments_in_block_to_match_guard_works() {
        check_assist(
            move_arm_cond_to_match_guard,
            r#"
fn main() {
    match 92 {
        x => {
            // Comment A
            if x > 10 {<|>
                // Comment B
                false
            }
        },
        _ => true
    }
}
"#,
            r#"
fn main() {
    match 92 {
        // Comment A
        // Comment B
        x if x > 10 => false,
        _ => true
    }
}
"#,
        );
    }
```
</details>

Or should those blocks not be "unwrapped" if there are comments?

Reply:
>How should comments in blocks be handled? Does the behavior of these tests make sense?

I think *somehow* is the right answer at this stage, we need more practical experience to figure out the best practicies here.